graph/src/components/store/mod.rs

```diff
@@ -137,6 +137,33 @@ pub struct EntityKey {
     pub causality_region: CausalityRegion,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct EntityDerived {
+    /// Name of the entity type.
+    pub entity_type: EntityType,
+
+    pub entity_field: Word,
+
+    /// ID of the individual entity.
+    pub entity_id: Word,
+
+    /// This is the causality region of the data source that created the entity.
+    ///
+    /// In the case of an entity lookup, this is the causality region of the data source that is
+    /// doing the lookup. So if the entity exists but was created on a different causality region,
+    /// the lookup will return empty.
+    pub causality_region: CausalityRegion,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum EntityMultiKey {
+    /// A filter that matches all entities of a given type.
+    Derived(EntityDerived),
+
+    /// A filter that matches a specific entity.
+    Single(EntityKey),
+}
```

Comment:
I haven't read through the changes in detail, but two questions come to mind from this:
1. How are you going to deal with aliasing? The same entity can now appear with as a `Single` and as (several) `Derived` in the cache. That will lead to correctness issues
2. The size of an `EntityMultiKey` is quite a bit bigger than an `EntityKey` which will increase the amount of memory the `EntityCache` takes or reduce the number of entities that can go into the cache, either of which are not desirable.

Reply:
1. I'm planning to keep in the `Derived` cache only an Array<String> containing the `ids` so the entity cache won't be duplicated. With that, I can cache the whole Entity as usual with it's own `EntityKey`. Currently, the only missing feature is to add/remove the id of a newly created/removed entities to keep the `Derived` cache updated.
2. The `EntityKey` is used in multiple places all over the code and changing it would be almost a complete refactor of the code. The safest way I found was to wrap it inside the `EntityMultiKey` enum and use it only in the `EntityCache`. I agree that `EntityMultiKey` is bigger than `EntityKey` and I'm open to suggestions.

Reply:
What I had in mind was what I outlined in the comment on #4004 : make the lookup understand how to chase down how the original entity and the entities you want to look up are related, and then treat the result as normal entities so that there's no changes needed to the `EntityCache`. Basically, the flow I have in mind is:
1. mapping calls `loadRelated`
2. internally transform that into the right query (which will depend on whether the field being looked up is derived or not, see [here](https://github.com/graphprotocol/graph-node/blob/master/docs/implementation/sql-query-generation.md) for how that's done in general)
3. run that query and get back a list of entities
4. insert them into the cache as if they had been looked up one-by-one with `load`
5. return to the mappings

Reply:
Ok, so essentially, the response of `loadRelated()` would not be cached, just call to the database loading the entities to the cache.

I mean, if I call `loadRelated()` twice it should query the database twice. That's way simpler than what I'm currently doing.

Reply:
Aah .. now I see where you're coming from. Yes, this shouldn't be cached since it will be too hard to track what changes will cause the response to change. When we save an entity, we have no idea if that will affect the list of related entities somewhere else. For example, in the data model from #4004,  you could just save a new `Token` entity (or modify an existing one) and set its `holder` to one for which you already used `loadRelated`. To keep the cache accurate, we would have to find all entities that could be affected by that and update the cached list of things. 

If it turns out that repeatedly calling `loadRelated` is an important use case, we can try and add that later.
For now, just having it so that each `loadRelated` call goes to the database and adds the individual entities to the cache without tracking that they were put there by `loadRelated` is good enough.

Reply:
Ok, I removed all the cache logic, now it's way simpler. You can continue to review the PR 😊